### PR TITLE
Improve TFD Installation logic

### PR DESCRIPTION
**Pull request by LipkeGu**
This Pr adds more flexibilty to the TFD Installer when installing content from TFD discs. by using index lookup. Closes https://github.com/OpenRA/OpenRA/issues/9082. 

also this Pr pays attention to:  https://github.com/OpenRA/OpenRA/pull/8170#issuecomment-129621166

### CODE REVIEW

**Comment by GraionDilach** [comment]
:-1:. You were told why supporting multiple languages at TFD is a bad idea.

**Comment by LipkeGu** [comment]
It changes only the behaver Oo how the files gets detected. it adds no language support

**Comment by Unit158** [comment]
Looks reasonable. Will test tomorrow.

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs [hunk 1]
```diff
@@ -404,6 +404,16 @@ public string DirectoryName(uint index)
 			return test;
 		}
 
+		public uint GetIndexFromFilename(string directory, string filename)
+		{
+			var path = directory + "\\" + filename;
```
Comment (Mailaender) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: I am not sure as this is inside a CAB file, but `\` is the Windows path separator. Hardcoding it will break everywhere else. I suggest you `Path.DirectorySeparatorChar` instead.

Comment (LipkeGu) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: https://gist.github.com/DSUK/e243103217bac918b175 yes they are inside... you say in this case give me the IDs of the files... not give me the files of the IDs.

**Comment by LipkeGu** [comment]
updated

File: mods/cnc/mod.yaml [hunk 2]
```diff
@@ -153,7 +153,8 @@ ContentInstaller:
 		.: conquer.mix, desert.mix, general.mix, scores.mix, sounds.mix, temperat.mix, winter.mix
 	ShippedSoundtracks: 4
 	MusicPackageMirrorList: http://www.openra.net/packages/cnc-music-mirrors.txt
-	InstallShieldCABFileIds: 66, 45, 42, 65, 68, 67, 71, 47, 49, 60, 75, 73, 53
+	InstallShieldCABDirectory: CnC
+	FiletoExtractFromCab: SPEECH.MIX, CONQUER.MIX:CRC32, CCLOCAL.MIX:CRC32, SOUNDS.MIX, TEMPICNH.MIX, TEMPERAT.MIX, TRANSIT.MIX, DESERT.MIX, GENERAL.MIX, SCORES.MIX, WINTER.MIX, UPDATEC.MIX
```
Comment (pchote) [inline mods/cnc/mod.yaml]: What is the purpose of these `CRC32`s?

Comment (LipkeGu) [inline mods/cnc/mod.yaml]: hash checks which was already there :(

Comment (LipkeGu) [inline mods/cnc/mod.yaml]: should i remove them?

File: mods/ra/mod.yaml [hunk 3]
```diff
@@ -152,8 +152,8 @@ ContentInstaller:
 		.: INSTALL/REDALERT.MIX
 	ShippedSoundtracks: 2
 	MusicPackageMirrorList: http://www.openra.net/packages/ra-music-mirrors.txt
-	InstallShieldCABFilePackageIds: 105
-	InstallShieldCABFileIds: 116
+	InstallShieldCABDirectory: Red Alert
+	FiletoExtractFromCab: redalert.mix:crc32, main.mix
```
Comment (Mailaender) [inline mods/ra/mod.yaml]: CRC32 encryption is TS/RA2 era mix files.

**Comment by LipkeGu** [comment]
updated

File: OpenRA.Mods.Common/InstallUtils.cs [hunk 4]
```diff
@@ -39,11 +39,8 @@ public class ContentInstaller : IGlobalModData
 		public readonly string MusicPackageMirrorList = null;
 		public readonly int ShippedSoundtracks = 0;
 
-		/// <summary> InstallShield .cab File Ids, used to extract Mod specific files </summary>
-		public readonly int[] InstallShieldCABFileIds = { };
-
-		/// <summary> InstallShield .cab File Ids, used to extract Mod specific archives and extract contents of ExtractFilesFromCD </summary>
-		public readonly string[] InstallShieldCABFilePackageIds = { };
+		public readonly string InstallShieldCABDirectory = null;
+		public readonly string[] FiletoExtractFromCab = { };
```
Comment (penev92) [inline OpenRA.Mods.Common/InstallUtils.cs]: Please keep with the `InstallShieldCAB*` naming convention set for these.
Also don't remove the `<summary>`-ies.

Oh, and the ` = null` is not necessary.

File: mods/ra/mod.yaml [hunk 5]
```diff
@@ -152,8 +152,8 @@ ContentInstaller:
 		.: INSTALL/REDALERT.MIX
 	ShippedSoundtracks: 2
 	MusicPackageMirrorList: http://www.openra.net/packages/ra-music-mirrors.txt
-	InstallShieldCABFilePackageIds: 105
-	InstallShieldCABFileIds: 116
+	InstallShieldCABDirectory: Red Alert
+	FiletoExtractFromCab: redalert.mix, main.mix
```
Comment (penev92) [inline mods/ra/mod.yaml]: I see 1 ID above turned into 2 names here. What's going on?

File: mods/cnc/mod.yaml [hunk 6]
```diff
@@ -153,7 +153,8 @@ ContentInstaller:
 		.: conquer.mix, desert.mix, general.mix, scores.mix, sounds.mix, temperat.mix, winter.mix
 	ShippedSoundtracks: 4
 	MusicPackageMirrorList: http://www.openra.net/packages/cnc-music-mirrors.txt
-	InstallShieldCABFileIds: 66, 45, 42, 65, 68, 67, 71, 47, 49, 60, 75, 73, 53
+	InstallShieldCABDirectory: CnC
+	FiletoExtractFromCab: SPEECH.MIX, CONQUER.MIX, CCLOCAL.MIX, SOUNDS.MIX, TEMPICNH.MIX, TEMPERAT.MIX, TRANSIT.MIX, DESERT.MIX, GENERAL.MIX, SCORES.MIX, WINTER.MIX, UPDATEC.MIX
```
Comment (penev92) [inline mods/cnc/mod.yaml]: 13 IDs turned into 12 file names here?

File: mods/ra/mod.yaml (hunk 5 above)
Comment (LipkeGu) [inline mods/ra/mod.yaml]: index lookup for this files

File: mods/cnc/mod.yaml (hunk 6 above)
Comment (LipkeGu) [inline mods/cnc/mod.yaml]: CabinetExtrator: Implement GetIndexByFilename

**Comment by LipkeGu** [comment]
has dependencies

**Comment by abcdefg30** [comment]
On what?

File: mods/cnc/mod.yaml (hunk 6 above)
Comment (DSUK) [inline mods/cnc/mod.yaml]: it's called MOVIES.MIX on TFD

File: mods/ra/mod.yaml (hunk 5 above)
Comment (penev92) [inline mods/ra/mod.yaml]: Which TFD disks?

File: mods/cnc/mod.yaml (hunk 6 above)
Comment (LipkeGu) [inline mods/cnc/mod.yaml]: @DSUK : thats why there is no movies.mix either movies***.mix Europe Disks has extracted vqa files

Comment (penev92) [inline mods/cnc/mod.yaml]: Which is exactly why people told you on GH and IRC that we can't support them.

**Comment by LipkeGu** [comment]
updated

**Comment by LipkeGu** [comment]
dependencies was related to RA2 but are now solved

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs [hunk 7]
```diff
@@ -404,9 +405,44 @@ public string DirectoryName(uint index)
 			return test;
 		}
 
+		public string GetFirstSubDirectory(string package)
+		{
+			var dir = "";
+			var delimeter = "{0}".F(Path.DirectorySeparatorChar);
+
+			foreach (var key in fileLookup.Keys)
+				if (key.StartsWith(package.ToLowerInvariant()) && key.Contains(delimeter))
```
Comment (pchote) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: Style nit: avoid hanging blocks.  Put braces around this please.

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs [hunk 8]
```diff
@@ -404,9 +405,44 @@ public string DirectoryName(uint index)
...
+			return dir;
+		}
+
+		public uint GetIndexByFilename(string directory, string filename)
+		{
+			var path = directory + Path.DirectorySeparatorChar + filename;
+
+			if (Exists(path.ToLowerInvariant()))
+				return fileLookup[path.ToLowerInvariant()];
+			else
```
Comment (pchote) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: Style nit: this `else` is unnecessary (you can unconditionally throw if it doesn't return).

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs [hunk 9]
```diff
@@ -91,44 +85,63 @@ void InstallTFD(string source)
...
+
+						dest = Platform.ResolvePath("^", "Content", Game.ModData.Manifest.Mod.Id, filename.ToLowerInvariant());
+						cabExtractor.ExtractFile(fileindex, dest);
 						progressBar.Percentage += installPercent;
 					}
 
-					var ArchivesToExtract = installData.InstallShieldCABFilePackageIds.Select(x => x.Split(':'));
+					progressBar.Indeterminate = true;
+
+					var ArchivesToExtract = installData.ExtractFromInstallShieldPackage.Select(x => x.Split(':'));
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs]: style nit: local variables shouldn't start with a capital letter.

File: mods/cnc/mod.yaml [hunk 10]
```diff
@@ -153,7 +153,8 @@ ContentInstaller:
 		.: conquer.mix, desert.mix, general.mix, scores.mix, sounds.mix, temperat.mix, winter.mix
 	ShippedSoundtracks: 4
 	MusicPackageMirrorList: http://www.openra.net/packages/cnc-music-mirrors.txt
-	InstallShieldCABFileIds: 66, 45, 42, 65, 68, 67, 71, 47, 49, 60, 75, 73, 53
+	InstallShieldPackageName: CnC
+	ExtractFromInstallShieldPackage: SPEECH.MIX, CONQUER.MIX, CCLOCAL.MIX, SOUNDS.MIX, TEMPICNH.MIX, TEMPERAT.MIX, TRANSIT.MIX, DESERT.MIX, GENERAL.MIX, SCORES.MIX, WINTER.MIX, UPDATEC.MIX
```
Comment (pchote) [inline mods/cnc/mod.yaml]: The case changes across mods are worrying.  Somebody with a case sensitive filesystem will need to test this before we merge.

File: mods/ts/mod.yaml [hunk 11]
```diff
@@ -190,18 +190,18 @@ LoadScreen: LogoStripeLoadScreen
...
-	TestFiles: cache.mix, conquer.mix, isosnow.mix, isotemp.mix, local.mix, sidec01.mix, sidec02.mix, sno.mix, snow.mix, sounds.mix, speech01.mix, tem.mix, temperat.mix
+	TestFiles: cache.mix, conquer.mix, isosnow.mix, isotemp.mix, local.mix, sidec01.mix, sidec02.mix, sno.mix, snow.mix, sounds.mix, speech01.mix, speech02.mix, tem.mix, temperat.mix
 	PackageMirrorList: http://www.openra.net/packages/ts-mirrors.txt
 	DiskTestFiles: MULTI.MIX, INSTALL/TIBSUN.MIX
 	CopyFilesFromCD:
 		.: INSTALL/TIBSUN.MIX, SCORES.MIX, MULTI.MIX
 	PackageToExtractFromCD: INSTALL/TIBSUN.MIX:CRC32
 	ExtractFilesFromCD:
-		.: cache.mix, conquer.mix, isosnow.mix, isotemp.mix, local.mix, sidec01.mix, sidec02.mix, sno.mix, snow.mix, sounds.mix, speech01.mix, tem.mix, temperat.mix
+		.: cache.mix, conquer.mix, isosnow.mix, isotemp.mix, local.mix, sidec01.mix, sidec02.mix, sno.mix, snow.mix, sounds.mix, speech01.mix, speech02.mix, tem.mix, temperat.mix
```
Comment (pchote) [inline mods/ts/mod.yaml]: These speech02 changes aren't related to the TFD changes, so should live in their own commit or (preferably, assuming they are fixing a bug) in their own PR.

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs [hunk 12]
```diff
@@ -404,9 +405,44 @@ public string DirectoryName(uint index)
 			return test;
 		}
 
+		public string GetFirstSubDirectory(string package)
```
Comment (pchote) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: What is the purpose of this method?  The implementation looks to be DIYing some path manipulation that the standard library should provide methods for.

Comment (penev92) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: >German disks has 13 cabinets. so i decided to check for data1.hdr only.

TLDR he's hacking it because German and non-german TFD contents is different.
Which, again, is the exact reason why people have been saying we shouldn't try to support them.

Comment (pchote) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: I'm deliberately ignoring the motivations because replacing the opaque file IDs with filenames is a good idea.

Comment (pchote) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: Ah, I see.  That should not be included in this PR.  Replacing the opaque and arbitrary file IDs with filenames is good, but we are not ready to support non-english disks OR RA2 right now.

**Comment by penev92** [comment]
OK let's make the purpose of this PR clear, because "Improve TFD Installation logic" and "adds more flxibility" aren't.

This should replace the file IDs in YAML with file names.
This should **not** add support for non-english disks.
This should **not** try to add support for RA2, if that includes any hacks in the code.

**Comment by GraionDilach** [comment]
@penev92: Except that it does _exactly_ what you listed, nothing.

File: mods/ts/mod.yaml (hunk 11 above)
Comment (LipkeGu) [inline mods/ts/mod.yaml]: no it was a problem in the first tfd PR :/ but nobody found the problem... so i added it here also :(

Comment (penev92) [inline mods/ts/mod.yaml]: Can you explain this? Because this line has nothing to do with installation from a TFD DVD, and was not touched by the PR that introduced support for TFD.

**Comment by LipkeGu** [comment]
@pchote i compared the us disk with the german disk and there are no differences. all filenames are the same. but only the folder names are different. They are Binary Compatible also mailaender sayed that (he is an german guy) and for years i play with german TS disks ...

The intentions to improve that Installer was not to add GERMAN support it was mainly an idea to add generic support for ALL TFD Disks. What Im not understandig is we say in the News.

"Added support for TFD disks"... that means for german users "support for all disks" we did not say "ONLY FOR US Disks" but now we have in the first PR 2 comments about that german disks will not be able to use because of the IDs" because we wrote "support for TFD disks".

When we did not support german disks TFD disks why we does we support german versions of RA,TS and CNC?!?! since years... ?!?! many peoples are playing with german disks and there is no problems on this site. binary compatible

In this case when we add support for TFD then we should not limit this support for US only instead we should support EVERY TFD disk ... not only for I2 types of disks. we have all our Legals and rules, but it should not be limited to this statement: they reason why we does not support german disks are: german has oil instead of blood so thats why we doesnt support german disks... Oo
that statement was written... in the issue where i posted the german IDs of the files...

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs (hunk 8 above)
Comment (LipkeGu) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: that else handles conditions like when files can not be found in the file table... it must return indexe i can not return 0 because zero stays for a another file in the lookup table

File: mods/cnc/mod.yaml (hunk 10 above)
Comment (LipkeGu) [inline mods/cnc/mod.yaml]: why we have to test this?! the filenames will always lowercase

https://github.com/LipkeGu/OpenRA/commit/54f878993b02c2d8e3ddcd98ba01dcd74734c056#diff-d7d7b22ed1d8f9159c5bc9aa83fdb4d0R437

File: mods/ts/mod.yaml (hunk 11 above)
Comment (LipkeGu) [inline mods/ts/mod.yaml]: you self sayed speech02 is missing on TS

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs (hunk 12 above)
Comment (LipkeGu) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: thats does not enable german support

**Comment by LipkeGu** [comment]
closing for now because there is no intention for this pr

**Comment by penev92** [comment]
Well OK, since we discussed it on IRC last night and figured out the misunderstanding I don't know why you'd close it before someone tested it, but OK.

**Comment by LipkeGu** [comment]
as dicussed in IRC

<penev> we keep the ID lookup by name (which is nice); we add a check if this is an english DVD; if not, we issue a warning along the lines of "non-english versions of TFD are not supported. Install at your own risk!"
<GraionDilach> that's acceptable```
```

**Comment by LipkeGu** [comment]
updated

**Comment by GraionDilach** [comment]
I can't find where it throws the warning. That change is still requested.

**Comment by LipkeGu** [comment]
That notice should be in the changelog

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs [hunk 13]
```diff
@@ -475,7 +492,7 @@ public Stream GetContentById(uint index)
 				reader.CopyTo(output);
 
 			if (output.Length != fileDes.ExpandedSize)
-				throw new Exception("Did not fully extract Expected = {0}, Got = {1}".F(fileDes.ExpandedSize, output.Length));
+				throw new IOException("Did not fully extract Expected = {0}, Got = {1}".F(fileDes.ExpandedSize, output.Length));
```
Comment (penev92) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: Fullstop or exclamation mark after "extract" please.

File: OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs [hunk 14]
```diff
@@ -500,7 +517,7 @@ public void GetContentById(uint index, Stream output)
...
```
Comment (penev92) [inline OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs]: And here.

**Comment by penev92** [comment]
No. We were over this on IRC. A warning in the changelog makes no difference.
In case @pchote and @Mailaender are OK with the suggestion, it can go in the game. Otherwise we can't take this PR, as discussed on IRC yesterday.

**Comment by LipkeGu** [comment]
yes then we need message box support in OpenRA =(

**Comment by GraionDilach** [comment]
You can add a label to the content copy process dialog where you can write the warning message in capslock.

**Comment by LipkeGu** [comment]
@GraionDilach : Only RA needs this warning: http://www.schnittberichte.com/schnittbericht.php?ID=3406

there is an missing movie ... tiberian sun has an changed movie and generals is completely changed :(

**Comment by LipkeGu** [comment]
updated

**Comment by Mailaender** [comment]
I don't get your obsessions with warnings about non-existing incompatibilities. Those assets are binary compatible. Even the Generals censorship could be undone with content patches only. Localisation is often done after the game has gone gold by non-technical people or it is outsourced. That is the reason we don't get into trouble here.

This unhardcodes ugly file offsets to make this work with any TFD disc. :+1: It makes the whole thing much more robust and sane. Can't test as I don't own the DVD myself.

**Comment by GraionDilach** [comment]
Binary compatible in Westwood terms means that the executable, art.ini, rules.ini, sound.ini and the map is the same. Everything else can differ.

The problem is that OpenRA does not react to the same for missing assets as the Westwood games. If an asset lost a frame, got removed because being unused, then the WW games will deal with it, but OpenRA throws a crash.

I still haven't seen any of you actually checking out to see if each file still exists within the localized mixes and that's just the German version. There might be more versions as well.

**Comment by Mailaender** [comment]
Please stop spreading that FUD. We have never received a crash report so far when someone used localized mix files even though there are many and even fan made ones out on the net. The worst thing that could happen here is spelling mistakes in icons and terrible voice actors.

**Comment by LipkeGu** [comment]
Changelog could be:

```
The compatibility with the Content Installer and "The First Decade" disks has been improved. Please keep in mind that the content in Europe can differ.
```

**Comment by Mailaender** [comment]
> yes then we need message box support in OpenRA =(

We already have `ConfirmationDialogs.PromptConfirmAction`.

**Comment by LipkeGu** [comment]
and depends label please because the asset code is moving....

**Comment by penev92** [comment]
Doesn't look like there will be any conflicts between the two PRs other than that moving `ContentInstaller` to a new file and this one changing 2 lines in it, which is a very simple rebase regardless of which one gets merged first.

**Comment by LipkeGu** [comment]
Otherwise we can move this together with your PR

**Comment by penev92** [comment]
Reasons against doing that:
 - My PR is big enough as it is and I wouldn't want to force more stuff on reviewers than is necessary.
 - I am not familiar with the changes in this PR, so I'm not comfortable with merging it in my own.
 - Both PRs are perfectly fine on their own as they have very little in common.

**Comment by Mailaender** [comment]
This just needs a +1 from people who have a TFD disc and can confirm that this works properly on every platform.

**Comment by LipkeGu** [comment]
On Windows it works so as far i can tell...

**Comment by LipkeGu** [comment]
updated and merged into one commit!

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs [hunk 15]
```diff
@@ -91,44 +85,74 @@ void InstallTFD(string source)
 			{
 				using (var cabExtractor = new InstallShieldCABExtractor(source))
 				{
-					var denom = installData.InstallShieldCABFileIds.Length;
+					var dest = "";
+					var files = installData.ExtractFromInstallShieldPackage.Select(x => x.Split(':'));
 					var extractFiles = installData.ExtractFilesFromCD;
+					var installPercent = 100 / files.Count();
+					var fileindex = (uint)0;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs]: `0u`.

**Comment by Mailaender** [comment]
Needs a rebase.

**Comment by pchote** [comment]
This pr actually does three different things, so it would be useful if you could separate this into three commits:

1. Code style changes in `InstallShieldCABExtractor`.
2. Replacing index &rarr; filename.
3. Fixing speech02.mix.

**Comment by steelphase** [comment]
It may be better as a separate PR, but this could fix #9480.

**Comment by penev92** [comment]
I think this is worth someone taking over at this point if @LipkeGu is missing.
Even if just for the ID -> file name change.

**Comment by DSUK** [comment]
tried installing this on my machine and got:
Exception of type `System.IO.DirectoryNotFoundException`: Could not find a part of the path "/home/james/.openra/content/ra/redalert.mix".

(Note: content has lowercase c here, There is Content with an uppercase C in my ~/.openra directory)

**Comment by LipkeGu** [comment]
Im not missing im looking at this at the Weekend :)

**Comment by abcdefg30** [comment]
No activity since a while, please reopen this when you come back to it.

**Comment by Mailaender** [comment]
Superseded by https://github.com/OpenRA/OpenRA/pull/11375.
